patches/api/0392-Add-LocalCoordinates-API.patch

```diff
@@ -0,0 +1,68 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: ChristopheG <61288881+chrisgdt@users.noreply.github.com>
+Date: Sat, 13 Aug 2022 13:57:03 +0200
+Subject: [PATCH] Add LocalCoordinates API
+
+
+diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
+index 840aaf9e8fc828b5a7ea02252038c6524680f2e0..41876a2cf3b02eaa33d838ed149e410b1b682951 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -903,6 +903,27 @@ public final class Bukkit {
+         return server.createExplorerMap(world, location, structureType, radius, findUnexplored);
+     }
+ 
++    // Paper start
++    /**
++     * Computes local coordinates according to the given location
++     * and coordinates.
++     * <br>
++     * Local coordinates are used in vanilla with <code>^x ^y ^z</code> and
```

Comment:
doesn't vanilla use tilde (~)?

Reply:
Those are for [relative coordinates](https://minecraft.fandom.com/wiki/Coordinates#Relative_world_coordinates), [local coordinates](https://minecraft.fandom.com/wiki/Coordinates#Local_coordinates) also take rotation into account.

Reply:
Yes it's just a little sad that this cannot be moved into the Location class due to nms needs

Reply:
Can't we just:
a) Copy the calculation logic into the location class?
or
b) Shove the NMS call into unsafe and call that from the Location class?

Reply:
Yes both will work but the unsafe class is meh...
Redoing the calculation is probably fine since it's just pure math, i think i prefer this over the unsafe use.

Reply:
If this has anything to do with Location, it should probably wait until the Position API is in. See the pr for that.

Reply:
Both ~ and ^ are used in vanilla, but we cannot mix them (we can use direct or relative coordinates but if we want to use local, the three components must be local coordinates).

The exact same calculation logic using API methods is the following (partially copied from NMS) : 
```
private static final float[] SIN = make(new float[65536], (sineTable) -> {
    for (int i = 0; i < sineTable.length; ++i) {
        sineTable[i] = (float)Math.sin((double)i * Math.PI * 2.0D / 65536.0D);
    }
});

private static <T> T make(T object, Consumer<T> initializer) {
    initializer.accept(object);
    return object;
}

private static float sin(float value) {
    return SIN[(int)(value * 10430.378F) & '\uffff'];
}

private static float cos(float value) {
    return SIN[(int)(value * 10430.378F + 16384.0F) & '\uffff'];
}

public static Location getLocalCoord(double left, double up, double forwards, Location origin) {
    float f = cos((origin.getYaw() + 90.0F) * ((float)Math.PI / 180F));
    float g = sin((origin.getYaw() + 90.0F) * ((float)Math.PI / 180F));
    float h = cos(-origin.getPitch() * ((float)Math.PI / 180F));
    float i = sin(-origin.getPitch() * ((float)Math.PI / 180F));
    float j = cos((-origin.getPitch() + 90.0F) * ((float)Math.PI / 180F));
    float k = sin((-origin.getPitch() + 90.0F) * ((float)Math.PI / 180F));

    Vector vec32 = new Vector(f * h, i, (double)(g * h));
    Vector vec33 = new Vector(f * j, k, (double)(g * j));
    Vector vec34 = vec32.clone().crossProduct(vec33).multiply(-1.0D);
    double d = vec32.getX() * forwards + vec33.getX() * up + vec34.getX() * left;
    double e = vec32.getY() * forwards + vec33.getY() * up + vec34.getY() * left;
    double l = vec32.getZ() * forwards + vec33.getZ() * up + vec34.getZ() * left;
    return new Location(origin.getWorld(), origin.getX() + d, origin.getY() + e, origin.getZ() + l);
}
```

Before discovering it, I had another code but it wasn't the same, especially when `left` is not zero. When I saw that my code wasn't the same local coordinates than the vanilla behavior, I created the PR.
```
Vector dirX = new Location(origin.getWorld(), 0, 0, 0, Location.normalizeYaw(origin.getYaw()-90), origin.getPitch()).getDirection().normalize();
Vector dirY = new Location(origin.getWorld(), 0, 0, 0, origin.getYaw(), origin.getPitch()-90).getDirection().normalize();
Vector dirZ = origin.getDirection().normalize();
return origin.add(dirX.multiply(left)).add(dirY.multiply(up)).add(dirZ.multiply(forwards));
```

I just wanted to link the actual behavior by having the same *sin* table (no idea if we can just replace sin and cos methods by Math.sin and Math.cos). I also though to set this in the Location class, but since it is located in the API, get a direct link to NMS wasn't a clear idea for me. I didn't add relative coordinates here because it is really easy for developers to just add an offset (actually, an API method like `Location#parse(String)` that takes a coordinate string, e.g. `~1 10 ~-4` or `^ ^1 ^5` would be great for everybody).

There is probably some better ways to add a method for local coordinates, so yes. The Position API is #7639 right ?

Reply:
Oh i have forgot that cos/sin wasn't exposed api side. So maybe use unsafe then...
For the difference between cos/sin table and the native java usage (if you see the impl of cos/sin you will see that use a native method that rely on some low level C method: if you are curious see: https://github.com/openjdk/jdk/blob/master/src/java.base/share/native/libfdlibm/s_cos.c) it's just an optimization that minecraft made for performance (it cache some values so the result will loss some accuracy). But i think this should return the exact same result as vanilla one so don't use native cos/sin.